target_chains/starknet/contracts/src/hash.cairo

```diff
@@ -0,0 +1,134 @@
+use super::reader::{Reader, ReaderImpl};
+use core::cmp::min;
+use core::integer::u128_byte_reverse;
+use pyth::util::{
+    ONE_SHIFT_160, UNEXPECTED_OVERFLOW, UNEXPECTED_ZERO, ONE_SHIFT_64, one_shift_left_bytes_u64,
+    u64_byte_reverse,
+};
+
+/// Allows to push data as big endian to a buffer and apply
+/// the keccak256 hash.
+#[derive(Drop, Debug)]
+pub struct Hasher {
+    // Inputs in little endian.
+    inputs_le: Array<u64>,
+    // Last pushed bytes in big endian.
+    last_be: u64,
+    // Number of filled bytes in `self.last_be`.
+    num_last_bytes: u8,
+}
+
+#[generate_trait]
+pub impl HasherImpl of HasherTrait {
+    /// Creates an empty hasher.
+    fn new() -> Hasher {
+        Hasher { inputs_le: array![], last_be: 0, num_last_bytes: 0 }
+    }
+
+    fn push_u8(ref self: Hasher, value: u8) {
+        self.push_to_last(value.into(), 1);
+    }
+    fn push_u16(ref self: Hasher, value: u16) {
+        self.push_num_bytes(value.into(), 2);
+    }
+    fn push_u32(ref self: Hasher, value: u32) {
+        self.push_num_bytes(value.into(), 4);
+    }
+    fn push_u64(ref self: Hasher, value: u64) {
+        self.push_num_bytes(value, 8);
```

Comment:
I am unsure that these will be used in the merkle tree in the future.

Reply:
Yeah, we'll only need `push_u8`, `push_u160` and `push_reader`, but I added the rest just for completeness.